src/compiler/utilities.ts

```diff
@@ -5781,13 +5781,13 @@ namespace ts {
      *  List of supported extensions in order of file resolution precedence.
      */
     export const supportedTSExtensions: readonly Extension[] = [Extension.Ts, Extension.Tsx, Extension.Dts];
-    export const supportedTSExtensionsWithJson: readonly Extension[] = [Extension.Ts, Extension.Tsx, Extension.Dts, Extension.Json];
+    export const supportedTSExtensionsWithJson: readonly Extension[] = [Extension.Ts, Extension.Tsx, Extension.Json, Extension.Dts];

```

Comment:
This isn't ideal I think... 

Consider scenario:: 
```ts
// moduleA.ts
import { x } from "./moduleB";

// moduleB.js
export const x = 10;
```

When you build first time files in your program with `allowJs` and `emitDeclarationsOnly` will be:
`moduleA.ts, moduleB.js`
Next time they will be 
`moduleA.ts, moduleB.d.ts and moduleB.js` since module resolution prefers up `d.ts` over `.js`

Reply:
That's what we do _today_. This PR fixes that, such that the .d.ts is never preferred if the js is present, assuming allowJs is on.

Reply:
What I mean is rootFiles will be `moduleA.ts and moduleB.js` but final list of files will contain .d.ts since that's what module resolution will resolve to? https://github.com/microsoft/TypeScript/blob/master/src/compiler/moduleNameResolver.ts#L1096

Reply:
Hm, yeah, the order we check extensions in module resolution likely needs to match the order wildcards use. I'll look at changing it to use the same source, ideally.